Reuse model binders

martincostello opened the pull request:
Reuse the same model binders for `CancellationToken`s and services, rather than allocating a new one for every request with such a parameter.

Both `CancellationTokenModelBinder` and `ServicesModelBinder` don't have any state, so it seems that reusing the same instance within the model binder providers ~(created lazily)~ would reduce the number of allocations for an application with a controller method similar to the example below:

```csharp
public async Task<IActionResult> GetForecast(
    string location,
    [FromServices] WeatherClient client,
    CancellationToken cancellationToken)
{
    var forecast = await client.GetForecastAsync(location, cancellationToken);
    return Json(forecast);
}
```

martincostello committed:
Reuse model binders

Reuse the same model binders for CancellationTokens and services, rather than allocating a new one for every request with such a parameter.

pranavkm commented on src/Mvc/Mvc.Core/src/ModelBinding/Binders/CancellationTokenModelBinderProvider.cs:
```diff
@@ -11,6 +11,8 @@ namespace Microsoft.AspNetCore.Mvc.ModelBinding.Binders
     /// </summary>
     public class CancellationTokenModelBinderProvider : IModelBinderProvider
     {
+        private CancellationTokenModelBinder _modelBinder;
```
```suggestion
        // CancellationTokenModelBinder  does not have any state. Re-use the same instance for binding.
        private readonly CancellationTokenModelBinder _modelBinder = new CancellationTokenModelBinder();
```

pranavkm replied:
Instantiating it up front seems fine. It's one additional instance per app, so I wouldn't be too worried about not allocating it. Same suggestion with the other type + the comment

martincostello committed:
Make fields readonly

Make the fields backing the binder once when initialized, rather than on first use.

Co-Authored-By: Pranav K <pranavkm@users.noreply.github.com>

pranavkm reviewed (approved):
Thanks!
